### PR TITLE
Add support for the TLS registry to the mailer extension

### PR DESCRIPTION
**Pull request by cescoffier**
It is now possible to use the TLS registry to configure the mailer extension. Note that it's recommended to use a named configuration in this case, to avoid conflicts.

### CODE REVIEW

File: extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/Mailers.java [hunk 1]
```diff
@@ -22,15 +22,9 @@
...
-import io.vertx.core.net.PfxOptions;
-import io.vertx.core.net.TrustOptions;
-import io.vertx.ext.mail.CanonicalizationAlgorithm;
-import io.vertx.ext.mail.DKIMSignOptions;
-import io.vertx.ext.mail.LoginOption;
-import io.vertx.ext.mail.MailClient;
-import io.vertx.ext.mail.StartTLSOptions;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.net.*;
+import io.vertx.ext.mail.*;
```
Comment (geoand) [inline extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/Mailers.java:27]: We generally don't like these :)

File: extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/Mailers.java [hunk 2]
```diff
@@ -60,7 +54,8 @@ public Mailers(Vertx vertx, io.vertx.mutiny.core.Vertx mutinyVertx, MailersRunti
         TlsConfiguration defaultTlsConfiguration = tlsRegistry.getDefault().orElse(null);
 
         if (mailerSupport.hasDefaultMailer) {
-            MailClient mailClient = createMailClient(vertx, mailersRuntimeConfig.defaultMailer, defaultTlsConfiguration);
+            MailClient mailClient = createMailClient(vertx, "<default>", mailersRuntimeConfig.defaultMailer,
```
Comment (geoand) [inline extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/Mailers.java:57]: Don't we have `"<default>"` as a constant somewhere?

File: extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailerRuntimeConfig.java [hunk 3]
```diff
@@ -70,17 +70,38 @@ public class MailerRuntimeConfig {
...
+     * The name of the TLS configuration to use.
+     * <p>
+     * If not set and the default TLS configuration is configured ({@code quarkus.tls.*}) then that will be used.
+     * If a name is configured, it uses the configuration from {@code quarkus.tls.<name>.*}
+     * If a name is configured, but no TLS configuration is found with that name then an error will be thrown.
+     * <p>
+     * If no TLS configuration is set, and {@code quarkus.tls.*} is not configured, then,
+     * `quarkus.mailer` SSL configuration will be used.
+     * <p>
+     * Important: This is only supported when using the Quarkus (Vert.x-based) gRPC client.
```
Comment (geoand) [inline extensions/mailer/runtime/src/main/java/io/quarkus/mailer/runtime/MailerRuntimeConfig.java:83]: I assume you don't want to mention gRPC

**Review by geoand — APPROVED** [review]

**Review by gsmet — APPROVED** [review]
Looks fine.

I added a small question. I'm not exactly sure what's best and our best option might be to gather some feedback from the field as to what is more practical.

So probably a good idea to merge and see how it goes.

File: docs/src/main/asciidoc/mailer-reference.adoc [hunk 4]
```diff
@@ -591,13 +591,38 @@ quarkus.mailer.port=465
...
+If your SMTP server uses a valid (trusted) certificate, you need to enable TLS explicitly (as you do not have to configure a trust store):
+
+[source, properties]
+----
+quarkus.tls.my-mailer.enable-tls=true
+quarkus.mailer.tls-configuration-name=my-mailer # Reference the named configuration
+----
+
+IMPORTANT: When using the mailer, using a _named_ configuration is recommended to avoid conflicts with other TLS configurations.
+Indeed, if a default TLS configuration is set, the mailer will automatically use it.
```
[the remark this answers is not in the document]
Comment (cescoffier) [reply docs/src/main/asciidoc/mailer-reference.adoc:624]: I totally agree. I'm not sure either. For http / gRPC servers it makes sense. For the clients, I have the feeling it's going to be more annoying than anything. 

So, now that I am not alone to think this, I think it's better to not use the default configuration for clients.

There is one annoying thing however: trust-all. The default trust all is used in many extensions (but not all...). I think we would need to keep this behavior (with a warning).

@geoand what do you think?

[the remark this answers is not in the document]
Comment (geoand) [reply docs/src/main/asciidoc/mailer-reference.adoc:624]: Yeah, when I was looking into OTel and REST Clients I encountered something similar.

So +1 for **not** using the `default` in clients and tell users to explicitly set a named one.

[the remark this answers is not in the document]
Comment (cescoffier) [reply docs/src/main/asciidoc/mailer-reference.adoc:624]: Alright, we have a deal. Let me modify the various issues I created with this new requirement. I will fix the mailer and redis ones.

[the remark this answers is not in the document]
Comment (geoand) [reply docs/src/main/asciidoc/mailer-reference.adoc:624]: 👌🏼

[the remark this answers is not in the document]
Comment (cescoffier) [reply docs/src/main/asciidoc/mailer-reference.adoc:624]: (might actually need a short ADR to explain the rationale - so we do not loosing this)

[the remark this answers is not in the document]
Comment (cescoffier) [reply docs/src/main/asciidoc/mailer-reference.adoc]: Fixed for the mailer.

**Review by gsmet — APPROVED** [review]
LGTM.

**Comment by gsmet** [comment]
CI not happy.

**Comment by cescoffier** [comment]
Yeah, will look at the CI.

**Comment by cescoffier** [comment]
@gsmet Well well well.... I may have forgotten a nasty detail about SMTP. Fortunately, we caught it.

Let me explain. SMTP connections are using TCP. 
You can have a plain TCP connection (unencrypted) - no problem, well, except for security.
You can have a TLS connection - no problem. This works with both the legacy configuration and the TLS registry

But, there is an in-between: `STARTTLS`. You establish a plain TCP connection and then send the `STARTTLS` command to start the TLS encryption. In this case, the legacy configuration is somewhat misleading:

```
quarkus.mailer.ssl=false # Plain connection
quarkus.mailer.trust-all=true # To trust the certificate sent by the server after the `STARTTLS` command
```

Sure, trusting all here is a security concern, but it worked before, and it does not now, as trust-all is ignored if TLS is disabled. I've restored this behavior when using the legacy configuration. I'm working on enabling it when using the TLS registry.

At the same time, I decided to extend and rework the mailer IT. They were using a very outdated fake SMTP server. Because we know MailPit is used as a dev service, I decided to use the same container (not the dev service) with various configurations to cover plain, `starttls`, and _full-tls_. This was not possible with the previous container. 

So, some work to do, but.... for the best.

**Pull request merged**